### PR TITLE
CDRIVER-5733 fix behavior of `bson_string_truncate`

### PR DESCRIPTION
**Pull request by kevinAlbs**
Fixes behavior of `bson_string_truncate` with large lengths. Verified with [this patch build](https://spruce.mongodb.com/version/66f186c226f05a00072fb0a4). See CDRIVER-5733 for details.

Also adds tests for behavior of `bson_string_truncate`. In the process of this fix, other unexpected behavior was observed: `bson_string_truncate` [documents](https://mongoc.org/libbson/1.28.0/bson_string_truncate.html) the truncated length "must be smaller or equal to the current length of the string.". However, `bson_string_truncate` does not reject greater lengths.
```c
bson_string_t *str = bson_string_new ("a");
bson_string_truncate (str, 2u); // Is not rejected.
ASSERT_CMPUINT32 (str->len, ==, 2u);
ASSERT_CMPUINT32 (str->alloc, ==, 4u);
bson_string_free (str, true);
```
This behavior is preserved (and now tested) for backwards compatibility. I expect there may be a possible use-case for calling `bson_string_truncate` with larger lengths to reserve space.

### CODE REVIEW

File: src/libbson/doc/bson_string_truncate.rst [hunk 1]
```diff
@@ -24,3 +24,4 @@ Truncates the string so that it is ``len`` bytes in length. This must be smaller
 
 A ``\0`` byte will be placed where the end of the string occurs.
 
+.. warning:: This function will abort if the length of the resulting string (including the NULL terminator) would exceed ``UINT32_MAX``.
```
Comment (eramongodb) [inline src/libbson/doc/bson_string_truncate.rst:27]:
```suggestion
.. warning:: The length of the resulting string (including the ``NULL`` terminator) MUST NOT exceed ``UINT32_MAX``.
```
Wording suggestion.

File: src/libbson/src/bson/bson-string.c [hunk 2]
```diff
@@ -307,19 +307,17 @@ void
...
-      alloc = 16;
+   BSON_ASSERT_PARAM (string);
+   if (len == string->len) {
+      return;
    }
-
-   if (!bson_is_power_of_two (alloc)) {
-      alloc = (uint32_t) bson_next_power_of_two ((size_t) alloc);
+   uint32_t needed = len;
+   BSON_ASSERT (needed <= UINT32_MAX - 1u);
```
Comment (eramongodb) [inline src/libbson/src/bson/bson-string.c:315]:
```suggestion
   BSON_ASSERT (needed < UINT32_MAX);
```

**Review by eramongodb — APPROVED** [review]
Minor feedback; otherwise, LGTM.

File: src/libbson/doc/bson_string_truncate.rst (hunk 1 above)
Comment (kevinAlbs) [reply src/libbson/doc/bson_string_truncate.rst:27]: Applied suggestion to other `bson_string_append*` functions with the same warning.

**Review by rcsanchez97 — APPROVED** [review]

**Review by rcsanchez97 — APPROVED** [review]

**Pull request merged**